pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -14,227 +14,202 @@
-    # We need a tolerance of around 1e-7 here in order to work with the case where U
-    # is specified with 8 decimal places.
-    if math.allclose(trace, 4, atol=1e-7) or math.allclose(trace, -4, atol=1e-7):
-        num_cnots = 0
-    # Case: 1 CNOT, the trace is 0
-    elif math.allclose(trace, 0.0, atol=1e-7):
-        num_cnots = 1
-    # Case: 2 CNOTs, the trace has only a real part
-    elif math.allclose(math.imag(trace), 0, atol=1e-7):
-        num_cnots = 2
 
-    return num_cnots
+def _decomposition_1_cnot(U, wires):
+    r"""If there is just one CNOT, we can write the circuit in the form.
+     -U- = -C--C--A-
+     -U- = -D--X--B-
 
+    To do this decomposition, first we find G, H in SO(4) such that
+        G (Edag V E) H = (Edag U E)
 
-def _su2su2_to_tensor_products(U):
-    r"""Given a matrix :math:`U = A \otimes B` in SU(2) x SU(2), extract the two SU(2)
-    operations A and B.
+    where V depends on the central CNOT gate, and both U, V are in SU(4). This
+    is done following the methods in https://arxiv.org/pdf/quant-ph/0308045.pdf.
 
-    This process has been described in detail in the Appendix of Coffey & Deiotte
-    https://link.springer.com/article/10.1007/s11128-009-0156-3
+    Once we find G and H, we can use the fact that E SO(4) Edag gives us
+    something in SU(2) x SU(2) to give A, B, C, D.
     """
+    # We will actually find a decomposition for the following circuit instead
+    # of the original U
+    # -U-SWAP- = -C--C-SWAP-B-
+    # -U-SWAP- = -D--X-SWAP-A-
+    # This ensures that the internal part of the decomposition has determinant 1.
+    swap_U = np.exp(1j * np.pi / 4) * math.dot(math.cast_like(SWAP, U), U)
 
-    # First, write A = [[a1, a2], [-a2*, a1*]], which we can do for any SU(2) element.
-    # Then, A \otimes B = [[a1 B, a2 B], [-a2*B, a1*B]] = [[C1, C2], [C3, C4]]
-    # where the Ci are 2x2 matrices.
-    C1 = U[0:2, 0:2]
-    C2 = U[0:2, 2:4]
-    C3 = U[2:4, 0:2]
-    C4 = U[2:4, 2:4]
-
-    # From the definition of A \otimes B, C1 C4^\dag = a1^2 I, so we can extract a1
-    C14 = math.dot(C1, math.conj(math.T(C4)))
-    a1 = math.sqrt(C14[0, 0])
+    # First let's compute gamma(u). For the one-CNOT case, uuT is always real.
+    u = math.dot(math.cast_like(Edag, U), math.dot(swap_U, math.cast_like(E, U)))
+    uuT = math.dot(u, math.T(u))
 
-    # Similarly, -C2 C3^\dag = a2^2 I, so we can extract a2
-    C23 = math.dot(C2, math.conj(math.T(C3)))
-    a2 = math.sqrt(-C23[0, 0])
+    # Since uuT is real, we can use eigh of its real part. eigh also orders the
+    # eigenvalues in ascending order.
+    _, p = math.linalg.eigh(qml.math.real(uuT))
 
-    # This gets us a1, a2 up to a sign. To resolve the sign, ensure that
-    # C1 C2^dag = a1 a2* I
-    C12 = math.dot(C1, math.conj(math.T(C2)))
+    # Fix the determinant if necessary so that p is in SO(4)
+    if math.linalg.det(p) < 0:
+        p = math.dot(p, math.cast_like(LAST_COL_NEG, p))
```

Comment:
This is not a suggestion atm, but more so my mind is still thinking of tracing.

I'm trying to think how this might be written to avoid the if statement 🤔

Reply:
surely there is a way of rewriting this so that it gives the correct answer dependent on the sign of `math.linalg.det(p)` 🤔

Reply:
Well... we could do:

```
LAST_COL_NEG = np.diag([1, 1, 1, math.sign(mat.linalg.det(p))])
```

But then we'd have to construct the matrix fresh every time.

Reply:
Ah, nice to know! I was staring at this with furrowed brows for a while

Reply:
Actually I'm gonna go with that, it just looks cleaner.